library/src/cache/patch_manager.rs

```diff
@@ -0,0 +1,883 @@
+    fn add_patch(&mut self, patch_number: usize, file_path: &Path) -> Result<()> {
+        if !file_path.exists() {
+            bail!("Patch file {} does not exist", file_path.display());
+        }
+
+        let patch_path = self.patch_artifact_path(patch_number);
+
+        std::fs::create_dir_all(self.patch_dir(patch_number))
+            .with_context(|| format!("create_dir_all failed for {}", patch_path.display()))?;
+
+        std::fs::rename(file_path, &patch_path)?;
+
+        let new_patch = PatchMetadata {
+            number: patch_number,
+            size: std::fs::metadata(&patch_path)?.len(),
+        };
+
+        // If a patch was never booted (next_boot_patch != last_booted_patch), we should delete
+        // it here before setting next_boot_patch to the new patch.
+        if let (Some(last_boot_patch), Some(next_boot_patch)) = (
+            self.patches_state.next_boot_patch,
+            self.patches_state.last_booted_patch,
+        ) {
+            if last_boot_patch.number != next_boot_patch.number {
+                if let Err(e) = self.delete_patch_artifacts(next_boot_patch.number) {
+                    error!(
+                        "Failed to delete patch artifacts for patch {}. Error: {}",
+                        patch_number, e
+                    );
+                }
+            }
+        }
+
+        self.patches_state.next_boot_patch = Some(new_patch);
+        self.patches_state.highest_seen_patch_number = self
+            .patches_state
+            .highest_seen_patch_number
+            .map(|highest_patch_number: usize| highest_patch_number.max(patch_number))
+            .or(Some(patch_number));
+        self.save_patches_state()
+    }
+
+    fn last_successfully_booted_patch(&self) -> Option<PatchInfo> {
+        self.patches_state
+            .last_booted_patch
+            .map(|patch| self.patch_info_for_number(patch.number))
+    }
+
+    fn next_boot_patch(&mut self) -> Option<PatchInfo> {
+        let next_boot_patch = match self.patches_state.next_boot_patch {
+            Some(patch) => patch,
+            None => return None,
+        };
+
+        if let Err(e) = self.validate_patch_is_bootable(&next_boot_patch) {
+            error!("Patch {} is not bootable: {}", next_boot_patch.number, e);
+
+            self.try_fall_back_to_last_booted_patch();
+
+            if let Err(e) = self.save_patches_state() {
+                error!("Failed to save patches state: {}", e);
+            }
+            return None;
+        }
+
+        self.patches_state
+            .next_boot_patch
+            .as_ref()
+            .map(|patch| self.patch_info_for_number(patch.number))
+    }
+
+    fn record_boot_success_for_patch(&mut self, patch_number: usize) -> Result<()> {
+        let next_boot_patch = self
+            .patches_state
+            .next_boot_patch
+            .context("No next_boot_patch")?;
+
+        if next_boot_patch.number != patch_number {
```

Comment:
Is next_boot_patch the current patch we're booting from?  Won't this fail if the updater thread somehow races ahead of dart initialization and has already installed a new patch before this is called?

Reply:
Just to make sure I understand (and to spell it out), the scenario we're worried about is:

1. The app launches using no patch.
2. The app downloads a new patch (call it `patch 1`, stores it as the `next_boot_patch`.
3. A new patch is published (call it `patch 2`).
4. The app restarts, booting from `patch 1` and downloading `patch 2`
5. `patch 2` finishes downloading and is marked `next_boot_patch` before `patch 1` finishes booting.
6. `record_boot_success_for_patch` is called for `patch 1`, but this method doesn't record it because `patch 2` is the next boot patch.

Is that correct?

Reply:
Correct.